File: constr_extractor/constraint.rb
# validate.rb in the old repo (formatchecker)
# how to write validate function: https://guides.rubyonrails.org/active_record_validations.html#validates-each
require "json"

class Constraint
  attr_accessor :class_name, :field_name
end

class InclusionConstraint < Constraint
  attr_accessor :values, :type

  def to_string
  end
end

class UniqueConstraint < Constraint
  attr_accessor :cond, :case_sensitive, :scope

  def initialize(classname, field, cond, case_sensitive, scope)
    @class_name = classname
    @field_name = field
    @cond = cond
    @case_sensitive = case_sensitive
    @scope = scope
  end

  def to_string
    {
      :constraint_type => "unique",
      :class_name => @class_name,
      :field => @field,
      :cond => @cond,
      :case_sensitive => case_sensitive,
      :scope => scope,
    }.to_json
  end
end

class PresenceConstraint < Constraint
  attr_accessor :cond

  def to_string
  end
end

class LengthConstraint < Constraint
  attr_accessor :min, :max

  def to_string
  end
end

class FormatConstraint < Constraint
  attr_accessor :format

  def to_string
  end
end

File: constropt/query_rewriter/constraint.py
class Constraint:
    def __init__(self, table, field) -> None:
        self.table = table
        self.field = field


class UniqueConstraint(Constraint):
    def __init__(self, table, field, scope=[], cond=None) -> None:
        if table == "principals":
            table = "users"

        self.table = table
        self.field = field
        self.scope = scope
        self.cond = cond

    def __str__(self) -> str:
        return "Unique, table: %s, field: %s, scope: %s, cond: %s" % (self.table, self.field, self.scope, self.cond)


class InclusionConstraint(Constraint):
    def __init__(self, table, field, values) -> None:
        if table == "principals":
            table = "users"
        super().__init__(table, field)
        self.value_list = values

    def __str__(self) -> str:
        return "Inclusion, table: %s, field: %s, value_list: %s" % (self.table, self.field, self.value_list)


class LengthConstraint(Constraint):
    def __init__(self, table, field, min, max) -> None:
        super().__init__(table, field)
        self.min = min
        self.max = max

    def __str__(self) -> str:
        return "Length, table: %s, field: %s, min: %s, max: %s" % (self.table, self.field, self.min, self.max)


class FormatConstraint(Constraint):
    def __init__(self, table, field, format) -> None:
        super().__init__(table, field)
        self.format = format

    def __str__(self) -> str:
        return "Format, table: %s, field: %s, format: %s" % (self.table, self.field, self.format)


class PresenceConstraint(Constraint):
    def __init__(self, table, field) -> None:
        super().__init__(table, field)

    def __str__(self) -> str:
        return "Presence, table: %s, field: %s" % (self.table, self.field)
